hackathon2023 constraintshub

richard-rogers opened the pull request:
## Description

Constraints Hub work for Hackathon 2023

## Changes

- Versioned store for config files
- [De]serialize constraint suite (and eventually validators, etc.) to/from the config store
- Workflow/UI for editing, reviewing, & managing constraint suites
- Autogenerate constraint suite to the workflow
- Profile writer that fakes evaluating constraints in WhyLabs & reports in profile columns for monitoring

richard-rogers committed:
Constraint-aware whylabs writer and ConfigHub API

richard-rogers committed:
work

FelipeAdachi committed:
read/write yaml files from/to factory constraints

richard-rogers committed:
local git config store

richard-rogers committed:
Merge branch 'dev/hackathon2023/constraintshub' of github.com:whylabs/whylogs into dev/hackathon2023/constraintshub

richard-rogers commented on python/whylogs/experimental/api/writer/cnwriter.py:
```diff
@@ -0,0 +1,22 @@
+    def write(self, file: Writable, **kwargs: Any) -> Tuple[bool, str]:
+        if isinstance(file, (DatasetProfile, DatasetProfileView)):
+            # check if {org,dataset}_id overridden in kwargs
+            constraints_str = ConfigHub(self._org_id, self._dataset_id, "constraints").get_latest()
```
I suppose we should check if the constraint suite is available :)

richard-rogers commented on python/tests/experimental/api/constraints/test_constraints_yaml.py:
```diff
@@ -0,0 +1,97 @@
+yaml_string = """\
+    
+id: "model-23"
+version: 1
```
What is this the version of? If it's the constraint suite, I've used semantic versioning so it would be 1.0.0.

FelipeAdachi replied:
that's just a placeholder at the moment, for now I focused only on constraints yaml.

richard-rogers commented on python/tests/experimental/api/constraints/test_constraints_yaml.py:
```diff
@@ -0,0 +1,97 @@
+    
+id: "model-23"
+version: 1
+hash: abcabc231 # maybe from git?
```
I'm not exposing hashes in the `ConfigStore` API, just semvers.

FelipeAdachi replied:
likewise, just placeholders, let's discuss what's needed and what's not

richard-rogers replied:
The `ConfigStore` is keyed with org id, model id, and kind of config (constraints, validators, etc). Those might be good meta data. There's also a commit message, though that's not currently fetchable in the API. I can add it if we want to include it here.

FelipeAdachi replied:
We can decide if the commit message is required or not, but I think it'd be important for the user to show his reasoning on why he's proposing the changes

richard-rogers commented on python/tests/experimental/api/constraints/test_constraints_yaml.py:
```diff
@@ -0,0 +1,97 @@
+            constraints=rehydrated_constraints, output_path=os.path.join(temp_dir, output_yaml_name)
+        )
+        constraints_out = read_constraints_from_yaml(os.path.join(temp_dir, output_yaml_name))
+        assert len(constraints) == len(constraints_out)
```
Could check `example.yaml` == `example2.yaml` by lines or file hashes?

FelipeAdachi replied:
I'm doing that on the test for yaml strings!

richard-rogers commented on python/whylogs/core/constraints/metric_constraints.py:
```diff
@@ -17,6 +17,13 @@
 ReportResult = namedtuple("ReportResult", "name passed failed summary")
 
 
+class ConstraintsParams(TypedDict):
```
This should probably be a `dataclass`

richard-rogers commented on python/whylogs/core/constraints/metric_constraints.py:
```diff
@@ -17,6 +17,13 @@
 ReportResult = namedtuple("ReportResult", "name passed failed summary")
 
 
+class ConstraintsParams(TypedDict):
```
Some of these seem redundant with the `MetricConstraint` itself?

FelipeAdachi replied:
You're right!
In latest commit, I tried to reduce redundancy by keeping only "factory" and specific parameters of factory constraints in ConstraintsParam. The other ones: "name", "column_name", "metric", I'm getting from the MetricConstraint ( and MetricSelector) itself.

richard-rogers commented on python/whylogs/experimental/api/constraints/__init__.py:
```diff
@@ -0,0 +1,69 @@
+
+
+def assemble_constraint(constraint_dict: dict) -> Optional[TypedDict]:
+    constraint_name = None
```
I think we could handle prefix expression constraints with something like
```
if "expression" in constraint_dict:  # alternative to factory key
  condition = PrefixCondition(constraint_dict.get("expression"))
  ...
elif "factory" in constraint_dict:
  # as below
```

For hackathon scope, just have the UI be a text entry field where we enter prefix expressions. For production, we can make a better expression builder UI that uses prefix internally.

FelipeAdachi replied:
If `expression` is present, constraint will be assembled using it, provided it has a name.

richard-rogers commented on python/whylogs/experimental/api/constraints/__init__.py:
```diff
@@ -0,0 +1,69 @@
+    constraints_list = []
+    for column_name, column_constraints in constraints.column_constraints.items():
+        for constraint_name, constraint in column_constraints.items():
+            if constraint._params:
```
Here check if the constraint callable is an instance of `PrefixCondition` and serialize with `- expression: constraint.condition._expression` instead of `factory: ...`

FelipeAdachi replied:
done!

richard-rogers committed:
use temp dir for test repos

richard-rogers commented on python/whylogs/experimental/extras/confighub.py:
```diff
@@ -0,0 +1,130 @@
+
+    @abstractmethod
+    def propose_version(self, config: str, new_version: Version, note: Optional[str] = None) -> None:
+        """
```
I think as currently written this can only be called once for a new version. Should I allow multiple calls to update the new version until it's committed (or dropped)? Should there be a separate `update_version(self, config, new_version, note)` method for updating an existing new version until it's committed (or dropped)?

richard-rogers replied:
Actually this might work for multiple calls.

richard-rogers commented on python/whylogs/experimental/extras/confighub.py:
```diff
@@ -0,0 +1,130 @@
+
+logger = logging.getLogger(__name__)
+
+
```
This API might not be great if we expect a lot of concurrent editing. Is that a use case we need to support?

FelipeAdachi replied:
I think that's probably ok for now

richard-rogers committed:
add ConfigStore::exists() and ConfigStore::create()

richard-rogers commented on python/whylogs/experimental/extras/confighub.py:
```diff
@@ -0,0 +1,147 @@
+class LocalGitConfigStore(ConfigStore):
+    def _run_commands(self, commands: List[str]) -> None:
+        for command in commands:
+            output = os.popen(f"cd {self._path}; {command}")
```
I should use the fancier `subprocess.Popen()` so I can capture (and ignore) the `stderr` output. It's annoying though.

richard-rogers replied:
I'm not sure how well this implementation will work on non-POSIX environments. For hackathon I don't care :)

richard-rogers committed:
test autocreate

richard-rogers commented on python/tests/experimental/api/constraints/test_constraints_yaml.py:
```diff
@@ -0,0 +1,97 @@
+        # Write the YAML string to a file
+        with open(yaml_path, "w") as file:
+            file.write(yaml_string)
+        constraints = read_constraints_from_yaml(os.path.join(temp_dir, input_yaml_name))
```
We need a `read_constraints_from_yaml_str(yaml: str)` for where the yaml has already been read from the `ConfigStore`.

FelipeAdachi replied:
done!

richard-rogers commented on python/tests/experimental/api/constraints/test_constraints_yaml.py:
```diff
@@ -0,0 +1,97 @@
+        builder = ConstraintsBuilder(dataset_profile_view=reference_profile_view)
+        builder.add_constraints(constraints)
+        rehydrated_constraints = builder.build()
+        write_constraints_to_yaml(
```
Likewise, constraints to yaml string that can be written by `ConfigStore`.

richard-rogers committed:
updated cnwriter

FelipeAdachi committed:
reduce redundancy between constraintparams and MetricConstraint
adding expression as parameter
adding reading/writing from/to yaml string in addition to file paths

FelipeAdachi committed:
add a few more constraints

richard-rogers commented on python/whylogs/experimental/api/writer/cnwriter.py:
```diff
@@ -0,0 +1,49 @@
+            and self._cn_store.exists()
+        ):
+            constraints_str = self._cn_store.get_latest()
+            constraints = read_constraints_from_yaml(constraints_str)
```
```suggestion
            constraints = read_constraints_from_yaml(input_str=constraints_str)
```

richard-rogers committed:
Update python/whylogs/experimental/api/writer/cnwriter.py

FelipeAdachi committed:
wip: frequent items

richard-rogers commented on python/whylogs/experimental/api/constraints/__init__.py:
```diff
@@ -0,0 +1,120 @@
+        return None
+
+
+class ConstraintTranslator:
```
If this class has no state I'd make the methods bare functions.

richard-rogers committed:
use ConstraintTranslator instance

richard-rogers committed:
add column name to ReportResult

richard-rogers committed:
pre-commit

FelipeAdachi committed:
endpoints implementation (#1225)

## Description

This repository needs <!-- Add 1-2 sentences explaining the purpose of
this PR. -->

The commits in this pull request will <!-- Summarize PR. -->

## Changes

- Imperative commit or change title (commit ID) <!-- Add commit ID and
[GitHub will
autolink](https://help.github.com/en/github/writing-on-github/autolinked-references-and-urls).
-->
  - Add some bullet points to explain the change
- Next commit or change (commit ID)

## Related

Relates to organization/repo#number

<!-- Reference related commits, issues and pull requests. Type `#` and
select from the list. -->

Closes organization/repo#number

<!-- List issues for GitHub to automatically close after merge to
default branch. Type `#` and select from the list. See the [GitHub docs
on linking PRs to
issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue)
for more. -->

- [ ] I have reviewed the [Guidelines for Contributing](CONTRIBUTING.md)
and the [Code of Conduct](CODE_OF_CONDUCT.md).

---------

Co-authored-by: murilommen <murilommen@gmail.com>

richard-rogers committed:
add constraint report to profile

richard-rogers committed:
Merge branch 'dev/hackathon2023/constraintshub' of github.com:whylabs/whylogs into dev/hackathon2023/constraintshub

richard-rogers commented on python/whylogs/experimental/api/writer/cnwriter.py:
```diff
@@ -0,0 +1,86 @@
+            report = builder.build().generate_constraints_report()
+            uncompounded_report = _uncompound_report(report)
+            report_view = why.log(row=uncompounded_report).profile().view()
+            file._columns.update(report_view._columns)
```
Views are meant to be read only. We should probably clone the original profile and insert the report columns into the clone...

FelipeAdachi replied:
wouldn't merging them work?
but for the hackathon I think this is fine either way!

richard-rogers replied:
The code works, but the side effect of modifying a read-only profile is bad :(  It's easy enough to do the clone & modify, but it can wait 'til after the demo :)

FelipeAdachi committed:
Merge branch 'dev/hackathon2023/constraintshub' of https://github.com/whylabs/whylogs into dev/hackathon2023/constraintshub

# Conflicts:
#	python/whylogs/experimental/api/constraints/__init__.py

FelipeAdachi committed:
wrong merge resolution

FelipeAdachi committed:
get latest

FelipeAdachi committed:
push_constraints

FelipeAdachi committed:
get parameters

richard-rogers commented on python/whylogs/experimental/constraints_hub/backend/routers.py:
```diff
@@ -0,0 +1,169 @@
+    storage_folder_name = "constraints_storage"
+    local_storage_folder = os.path.join(os.getcwd(), storage_folder_name)
+    cs = LocalGitConfigStore(org_id, dataset_id, "constraints", repo_path=local_storage_folder)
+    cs.create()
```
```suggestion
    if not cs.exists():
        cs.create()
```

richard-rogers commented on python/whylogs/experimental/constraints_hub/backend/routers.py:
```diff
@@ -0,0 +1,169 @@
+@router.get("/latest_version")
+def get_latest_version() -> None:
+    latest_yaml = cs.get_latest()
+    return {"constraints_yaml": str(latest_yaml)}
```
```suggestion
    return {"constraints_yaml": latest_yaml}
```
I think the `str()` is redundant?

richard-rogers commented on python/whylogs/experimental/constraints_hub/backend/routers.py:
```diff
@@ -0,0 +1,169 @@
+    yaml_string = yaml_constraint.constraints_yaml
+    constraints = translator.read_constraints_from_yaml(input_str=yaml_string)
+    cur_ver = cs.get_version_of_latest()
+    new_ver = cur_ver.next_major()
```
Maybe a UI widget/parameter to determine which version component gets bumped?

richard-rogers commented on python/tests/experimental/api/constraints/test_constraints_yaml.py:
```diff
@@ -0,0 +1,140 @@
+  factory: is_non_negative
+  metric: distribution
+  name: weight is non negative
+- expression: and <= animal:cardinality/lower_1 :animal:counts/n <= :animal:counts/n
```
```suggestion
- expression: and <= :animal:cardinality/lower_1 :animal:counts/n <= :animal:counts/n
```

richard-rogers committed:
typos

richard-rogers committed:
Merge branch 'dev/hackathon2023/constraintshub' of github.com:whylabs/whylogs into dev/hackathon2023/constraintshub

FelipeAdachi commented on python/whylogs/core/constraints/metric_constraints.py:
```diff
@@ -500,26 +518,26 @@ def validate(self, profile_view: Optional[DatasetProfileView] = None) -> bool:
 
     @deprecated(message="Please use generate_constraints_report()")
-    def report(self, profile_view: Optional[DatasetProfileView] = None) -> List[Tuple[str, int, int]]:
+    def report(self, profile_view: Optional[DatasetProfileView] = None) -> List[Tuple[Optional[str], str, int, int]]:
```
Also a thought for post-hackathon:
this might break report access to some users, but then again, it's a deprecated function, so maybe it' ok. or would it be better to not carry those changes into report(), and leave it only for generate_constraints_report()?

FelipeAdachi committed:
change yaml to json

FelipeAdachi committed:
Merge branch 'dev/hackathon2023/constraintshub' of https://github.com/whylabs/whylogs into dev/hackathon2023/constraintshub

FelipeAdachi committed:
adding more constraints parameters to get with backend's get_params

FelipeAdachi committed:
ref profile endpoints, fix on push_constraints

FelipeAdachi committed:
fix get latest

FelipeAdachi committed:
account for extra param column name in constraints report viz

FelipeAdachi committed:
fix report api

murilommen committed:
Constraints hub UI (#1226)

## Description

UI Functionalities for constraints Hub
